### DISCUSSION TITLE
Sanitize request parameters via filters

### DISCUSSION THREAD

**Discussion by visheshruparelia** [micronaut-projects/micronaut-core#9561] [Q&A]
Hi, taking an example of a GET request with URI parameters (say a URI like `GET /maps?name=unitedstates%20`) I want to add a filter which sanitizes the value of the `name` parameter like `String::trim`. 
I tried writing a filter like this:
```
@Filter(value = {V1_BASE_URL + "/**"})
public class ManagementServerFilter implements HttpServerFilter {

    static class FilteredRequest extends HttpRequestWrapper {
        FilteredRequest(HttpRequest httpRequest) {
            super(httpRequest);
        }

        @Override
        public HttpParameters getParameters() {
            Iterator<Map.Entry<String, List<String>>> itr = super.getParameters().iterator();
            while(itr.hasNext()) {
                Map.Entry<String, List<String>> entry = itr.next();
                entry.setValue(entry.getValue().stream().map(String::trim).collect(Collectors.toList()));
            }
            return super.getParameters();
        }
    }

    public Publisher<MutableHttpResponse<?>> doFilter(HttpRequest<?> request,
                                                      ServerFilterChain chain) {
        return Flowable.just(true)
                .switchMap(aBoolean -> chain.proceed(new FilteredRequest(request)));
    }
}
```

However, it doesnot seem to do the trick. Am I missing something here?

Also, what approach do I take for POST requests?
Thanks!

**Comment by graemerocher** [comment]
call `request.mutate()` and modify that

**Reply by visheshruparelia** [reply]
Thanks @graemerocher for the prompt response!

I tried this :
```
public Publisher<MutableHttpResponse<?>> doFilter(HttpRequest<?> request,
                                                      ServerFilterChain chain) {
        HttpParameters originalParams = request.getParameters();
        MutableHttpRequest mutableHttpRequest = request.mutate();
        MutableHttpParameters mutableHttpParameters = mutableHttpRequest.getParameters();
        originalParams.asMap().forEach((key, value) -> {
            mutableHttpParameters.add(key, value.stream().map(String::trim).collect(Collectors.toList()));
        });
        return Flowable.just(true)
                .switchMap(aBoolean -> chain.proceed(mutableHttpRequest));
    }
```
However, what this does is adds the new trimmed value instead of replacing it. 
For example, if `originalParams.asMap()` returns `{name=["unitedstates  "]}`, after I do `mutableHttpParameters.add()`, the output of `mutableHttpParameters.asMap()` returns `{name=["unitedstates  ", "unitedstates"]}`. This means that in my controller I am still getting `unitedstates  ` as input instead of `unitedstates`.

Am I missing something here?

**Reply by graemerocher** [reply]
Hm yeah good point. Generally this kind of normalisation and trimming is done in the binding implementation such as when you bind `@QueryValue("name") String name` rather than in a filter, but it is true there is a missing API to mutate parameters that is present in the `MutableHttpHeaders` API

**Reply by visheshruparelia** [reply]
@graemerocher the reason I am not doing it in controller but in filter is because if I do it in controller, for every endpoint, I will have to add `.trim()` method before handling the request which adds a lot of code duplication and will become difficult to manage. 
I was using RequestWrapper (code added in the original post) based on this stackoverflow answer: https://stackoverflow.com/a/1413145 but it does not work (any idea why?)
Is there any other approach which I can use to achieve this without code duplication? 
Please advise.

**Reply by graemerocher** [reply]
if you do it in an argument binder like https://github.com/micronaut-projects/micronaut-core/blob/4.0.x/http/src/main/java/io/micronaut/http/bind/binders/QueryValueArgumentBinder.java you won't have to do it for every endpoint. That said feel free to raise an issue to mutate the parameters in a filter.

**Comment by yawkat** [comment]
Using micronaut http 4 `@RequestFilter` you can accept a MutableHttpRequest as the parameter instead.

However I would recommend against sanitizing in a filter, because it is error-prone. If your logic does not match that of the argument binder exactly, it may be possible for an attacker to bypass your sanitization and pass an unsafe value to the controller. For example, if you forget the explicit `@QueryValue` annotation on the parameter, the parameter may also be bound from a multipart form body, which you did not sanitize.
